### ISSUE TITLE
Index: Filesystem watcher for auto indexing independent from WebDAV

### ISSUE THREAD

**Issue by graciousgrey** [photoprism/photoprism#2608]
**As a user, I want newly added or changed files to be indexed automatically, even if I don't use WebDAV.**

**Challenges:**
- Since our users use a variety of file systems and operating systems, we will not release something that has only been tested with the default file system on Linux, as this could lead to a huge number of support requests that we cannot handle.
- In practice, a single "photo" may consist of 3 or more files, some of which are created hours later - e.g. manually in Photoshop - and may have a completely different size or use (such as JPEG, RAW and XMP, which is only metadata and not an image that could be indexed on its own).
- This makes choosing a strategy that works for "everyone" difficult, even though each individual user may say it's "common sense".

**Acceptance Criteria:**
- [ ] The filesystem watcher MUST work on wide variety of file systems and operating systems
- [ ] The filesystem watcher MUST work for huge libraries (millions of files) 

**Research Needed:**
- [ ] Create a matrix of operating system and file system combinations, including any restrictions such as the number of files/folders supported.
- [ ] Under what conditions can file system watching fail? What happens when the user runs PhotoPrism in Docker running in a VM on Windows that has a network file system mounted?

**Comment by srett** [comment]
For the watcher, it might make sense to look at Syncthing. Also written in Go, handling huge directory trees just fine, and works across Linux/Windows/MacOS, probably more. It also has a fallback option to manually scan the FS every X minutes, because–as you also already mention–there will always be setups where it doesn't work, e.g. with some networked file systems. In the end you probably cannot reliably detect in advance if the watcher will work reliably. An accompanying disclaimer message next to that option's checkbox should suffice imo.
You could also display a warning message next time a scheduled/manual scan runs and you discover new files that the watcher should have found but didn't. But that in turn might lead to false positives under certain conditions, so I doubt it's worth the effort.

**Comment by e-dant** [comment]
You can use my library: https://github.com/e-dant/watcher

I'd be happy to help work it in with you.

**Comment by e-dant** [comment]
If we were to integrate my library fully, we'd create an extern block around the (only three) important symbols.

The FFI is not strictly necessary. You could just leave the program running (separately, in its own process group) and redirect its standard output into some buffer from which you read and parse the filesystem events (as JSON; their default output format).

For very large filesystem trees with very frequent events, this approach might not cut it. I'm not sure. I haven't tested this approach thoroughly. I can say, from empirical usage and our tests, that it chunks through and outputs millions of events on average hardware without breaking much of a sweat.

I started this project with the intention of it being a library as much as a program.

Still, I'm not sure this is the right library for you. Here's what I'm concerned about:
- my library is not mature
- you have many users. They will discover yet-unknown bugs. 
- there seem to be perfectly good filesystem watchers in Go. Syncthing (not the sync part) seems like a perfectly good option -- and it's much more mature.

Still, I've designed the `Watcher` to be as user-friendly, dependency-minimal (none so far!) and efficient as possible. There should be no fuss; it should just work. 

If the benefits of my library outweigh the caveats, you should use my library.

**Comment by lastzero** [comment]
Many of our users have large NTFS libraries mounted inside our Docker image via WSL2 / Hyper-V. Would a regular Linux filesystem watcher work with this at all and if so, could there be performance issues?

See https://docs.photoprism.app/getting-started/troubleshooting/windows/

**Comment by e-dant** [comment]
They would require superuser permissions and I would need to revive the `fanotify`-based adapter to be confident in that.

`inotify`-based adapters will not work, and the naive fallback adapter would not be performant enough.

I'm assuming permissions won't be an issue since this is running in a container.

Could you give me more details on how your user's containers are generally set up? Do you mount directories, run as privileged, use the same user group and id?

**Comment by lastzero** [comment]
Not aware of any UID mapping between Linux and Windows, so I assume the container would be running as root by default (it's a virtual machine anyway, from what I know). Users would then mount the folder(s) they want to index, either based on NTFS or ExFAT.

**Comment by e-dant** [comment]
I don't see why this shouldn't work with `fanotify` and some sanity checks.

I can estimate between 1 week and 1 month before I have this ready and running with tests.

`fanotify` is on my TODO list for the project, anyway. The adapter works; merging it is a matter of coming up with tests and ensuring it's only selected when the user's ID is 0.

**Comment by bigntallmike** [comment]
Seems to me that linking against fanotify should be sufficient. That said, if photoprism could be told about added/changed files externally, perhaps an external helper could run instead, for random example:
```
inotifywait -m /mnt/photos/* -e moved_to, close_write, modify |
    while read path action file; do
        photoprism --update "$path/$file"
```
Alternatively [fswatch](https://github.com/emcrisostomo/fswatch) is slightly more portable if that matters.

**Comment by srett** [comment]
@bigntallmike That's already possible. Not at home currently, but there is something like `photoprism index <path>` which you can invoke via `docker[-compose] exec`

**Comment by e-dant** [comment]
> Seems to me that linking against fanotify should be sufficient.

How do you mean?

> That said, if photoprism could be told about added/changed files externally, perhaps an external helper could run instead, for random example:
> 
> ```
> 
> inotifywait -m /mnt/photos/* -e moved_to, close_write, modify |
> 
>     while read path action file; do
> 
>         photoprism --update "$path/$file"
> 
> ```

`inotifywait` may be sufficient, but I'm skeptical. `inotify` explicitly doesn't support mounted volumes, and I'm not sure how that will look inside containers. It has other, more minor issues, like accuracy and possible overflows on very large directory trees. Those are detailed [bugs] section of its man page. 

> Alternatively [fswatch](https://github.com/emcrisostomo/fswatch) is slightly more portable if that matters.

`fswatch` seemed like a reasonable program to me in what it's capable of. It was also one of the reasons for me starting my project -- fswatch is over-complicated, lacking good unit tests, missing sanitizers and bulky to use as a library. YMMV, it is an established project with clearly defined capabilities.

**Comment by e-dant** [comment]
If you're still interested, the `fanotify` adapter on Watcher should be ready soon. It's awaiting some finishing touches [here](https://github.com/e-dant/watcher/blob/next/include/watcher/adapter/linux/fanotify/watch.hpp). I can likely get it into a release this weekend. We also have good unit tests and sanitizers for it (and the rest of the project) now.

Feel free to use another tool -- **whatever works best for you is best**.

But I'm happy to help get Watcher into use :)

**Comment by e-dant** [comment]
I may also fiddle with starting up some containers on Windows like so:

(From https://docs.photoprism.app/getting-started/docker)

```
docker run -d \
  --name photoprism \
  --security-opt seccomp=unconfined \
  --security-opt apparmor=unconfined \
  -p 2342:2342 \
  -e PHOTOPRISM_UPLOAD_NSFW="true" \
  -e PHOTOPRISM_ADMIN_PASSWORD="insecure" \
  -v /photoprism/storage \
  -v ~/Pictures:/photoprism/originals \
  photoprism/photoprism
```

On a fork with Watcher. With `fanotify`, the mounts to `photoprism/storage` and `~/Pictures:/photoprism/original` should not be a problem.

**Comment by graciousgrey** [comment]
@e-dant Thank you! We will get back to you as soon as we can start working on this!

**Comment by bmmathe** [comment]
Would it be too much effort to just add another api call like /api/v1/import?  There is an index endpoint but creating one for import would solve a lot of these feature requests.  
My example for needing this is similar to others.  I have an auto-copy from my desktop via iCloud sync to my unraid PhotoPrism "import" folder.  I then want to kick off an "import" after the copy.

Currently I have a job setup to run every x minutes which works fine but it would be nice to kick off the import immediately and only when needed.

**Comment by lastzero** [comment]
👉 https://github.com/photoprism/photoprism/blob/develop/internal/api/import.go#L32

**Comment by sphinnix** [comment]
I'm a noob, but I find a very simple way of autoindexing my photoprism library, and I think it can work basically for all linux distributions at least. It's clearly a workaround but maybe it can help.

Basically I created a script in bash that reads the import folder dimensions and write it to a text file. Then I have a service that every minutes launches the very same script, that if notices that the dimensions have changed, overwrite the text file with the new dimensions and launches the indexing command. Very simple and effective.

**Comment by mfed3** [comment]
Tried this project and liked it but uninstalled and got rid of it once I found out there is no automatic library update feature. Would not use ever again until this is implemented.

**Comment by lastzero** [comment]
Thank you for your feedback! You can now let your library get rescanned in the background at regular intervals. See the release notes and the configuration option overview in our documentation.

**Comment by tannisroot** [comment]
> Tried this project and liked it but uninstalled and got rid of it once I found out there is no automatic library update feature. Would not use ever again until this is implemented.

Photoprism is a more classic photo album application, if what you are looking for is a phone photo sync app, Immich may be a more suitable application.

**Comment by mfed3** [comment]
> > Tried this project and liked it but uninstalled and got rid of it once I found out there is no automatic library update feature. Would not use ever again until this is implemented.
> 
> Photoprism is a more classic photo album application, if what you are looking for is a phone photo sync app, Immich may be a more suitable application.

No, I'm using Syncthing for synchronization of photos among all my devices. I just want my server to always have the latest photos that are already on the server to show up in the web app running on that machine rather than me having to do something manual or wait for hours for the photos to show up. Seems like it should be pretty simple, since my cell phone can do it with an open source app called Gallery, despite having over a hundred gigabytes of photos.

**Comment by srett** [comment]
> wait for hours for the photos to show up

You can set the automatic indexing to run however often you want; try 1 minute?

**Comment by rayzorben** [comment]
> > wait for hours for the photos to show up
> 
> You can set the automatic indexing to run however often you want; try 1 minute?

And how is that going to perform on folders with 100,000+ photos?

**Comment by rayzorben** [comment]
> Tried this project and liked it but uninstalled and got rid of it once I found out there is no automatic library update feature. Would not use ever again until this is implemented.

I have to agree, I understand there are concerns with different scenarios, that's what options/config is for. I am looking for a tool to watch a folder `import` and any file that gets placed in that folder is automatically moved to `originals`, renamed as I choose (YYYY-mm-dd HH:mm:ss), restructured into folders as I choose (YYYY-mm) and then indexed, face recognition etc.

I have 3 kids, my wife and I, and my mother-in-law all with uploads to Dropbox. I want to provide a single app all of us can view those photos and create albums, share pictures with other family, delete bad photos, etc. - their expectation is to take some photos on their phone, log into PhotoPrism 5 minutes later and manage those photos immediately.

**Comment by srett** [comment]
> > > wait for hours for the photos to show up
> > 
> > 
> > You can set the automatic indexing to run however often you want; try 1 minute?
> 
> And how is that going to perform on folders with 100,000+ photos?

I don't know. My use case doesn't include uploading 100k photos every minute. :>

**Comment by mfed3** [comment]
People commenting are completely missing the point, the issue is that the indexing should be EVENT driven not time driven. New file added = add it instantly. If any android phone photo app can do this and support tens of thousands of photos in the database, running on an ARM processor, you guys can do it. Until then, I'm out.

**Comment by srett** [comment]
EDIT: Disregard this, I've too many open tabs. :facepalm: 

(@mfed3 you want inotify support. Maybe thumbs-up the according issue if you haven't done so already. (#2608)
But then again there are a bunch of other similar projects around, maybe one of them has that feature.)

**Comment by rayzorben** [comment]
> > > > wait for hours for the photos to show up
> > > 
> > > 
> > > You can set the automatic indexing to run however often you want; try 1 minute?
> > 
> > 
> > And how is that going to perform on folders with 100,000+ photos?
> 
> I don't know. My use case doesn't include uploading 100k photos every minute. :>

And how is it going to know which files are new and changed without scanning the whole library

**Comment by e-dant** [comment]
FWIW There's another Go project which is getting Watcher up and running via cgo here: https://github.com/dunglas/frankenphp/pull/1013

(But you should use the watcher that's right for you, whatever that turns out to be.)

**Comment by srett** [comment]
> And how is it going to know which files are new and changed without scanning the whole library

Ah I see. This could indeed become a bottleneck. From a one-off test with my lib (48k photos), it takes 22 seconds (SSD), so every minute might eventually become a problem.
